app/gui/src/model/suggestion_database.rs

```diff
@@ -27,6 +28,165 @@ pub use example::Example;
 
 
 
+// ====================
+// === FreeformPath ===
+// ====================
+
+type PathSegment = ImString;
+
+/// A path entered by a User, not guaranteed to correspond to any existing entity. The path is
+/// formed of segments with no guaranteed semantic meaning.
+///
+/// A conversion from a string value is performed by splitting the string into segments separated
+/// by the [`ACCESS`] character.
+#[derive(Clone, Debug, Default)]
+pub struct FreeformPath {
+    segments: Vec<PathSegment>,
+}
+
+impl FreeformPath {
+    fn from_segments<I, S>(segments: I) -> Self
+    where
+        I: IntoIterator<Item = S>,
+        S: Into<PathSegment>, {
+        let segments = segments.into_iter().map(|s| s.into()).collect();
+        Self { segments }
+    }
+}
+
+impl From<&str> for FreeformPath {
+    fn from(path: &str) -> FreeformPath {
+        FreeformPath::from_segments(path.split(ACCESS))
+    }
+}
+
+impl From<String> for FreeformPath {
+    fn from(path: String) -> FreeformPath {
+        path.as_str().into()
+    }
+}
+
+impl From<&Entry> for FreeformPath {
+    fn from(entry: &Entry) -> FreeformPath {
+        match entry.kind {
```

Comment:
As I mentioned, I don't feel how "conversion of entry to path" works. IMO getting full qualified name should be just a method of Entry.